src/lib.rs

```diff
@@ -0,0 +1,1166 @@
+/// ```
+///
+///
+/// ## Storage backing type
+///
+/// `Vob`s default to using `usize` as a storage backing type. This is generally a substantial win
+/// over using smaller storage types if you use functions such as
+/// [`or()`](struct.Vob.html#method.or). In such cases, `usize` on a 64-bit machine is almost
+/// exactly twice as fast as using `u32`. If you only ever set and get individual bits, a smaller
+/// data type might be marginally more effective: for such use cases `u32` is around 1% faster than
+/// `usize` on a 64-bit machine. You can choose your own storage type with the
+/// [`new_with_storage_type()`](struct.Vob.html#method.new_with_storage_type) constructor. In
+/// general we recommend using the default `usize` backing storage unless you have rigorously
+/// benchmarked your particular use case and are sure that a different storage type is superior.
+///
+///
+/// ## Migrating from `Vec<bool>`
+///
+/// As far as possible, `Vob` is intended to have a superset of `Vec<bool>`'s interface, which
+/// should make porting most code fairly simple. However, `Vec<bool>` contains several functions
+/// which are not yet implemented in `Vob`: these are missing simply due to a lack of a current
+/// use-case rather than because of any fundamental incompatibilities.
+///
+/// There is one missing feature which, currently, is impossible to implement: assignment to an
+/// index. In other words one cannot currently express `v[0] = true` for a `Vob` `v`. Until
+/// [`IndexGet` / `IndexMove` and equivalents](https://github.com/rust-lang/rfcs/issues/997) are
+/// implemented in `rustc`, this restriction appears to be inevitable. Note that referencing by
+/// index works (though via a hack identical to that used in `BitVec`): one can write
+/// `println!("{}", v[0])` for a `Vob` `v`, for example.
+///
+///
+/// ## Migrating from `BitVec`
+///
+/// `Vob` is directly inspired by `BitVec`, but aims to provide an interface aligned to `Vec<bool>`
+/// Several functions in `BitVec` have different names in `Vob`, but porting is in general fairly
+/// simple. The main semantic difference is that `Vob`s [`clear()`](struct.Vob.html#method.clear)
+/// function empties the `Vob` of contents (i.e. sets its length to 0), whereas `BitVec`s function
+/// of the same name unsets all bits (keeping the length unchanged). `Vob`s equivalent is
+/// [`set_all(false)`](struct.Vob.html#method.set_all).
+#[derive(Clone)]
+pub struct Vob<T=usize> {
+    /// How many bits are stored in this Vob?
+    bits: usize,
+    /// The underlying storage. Since the storage consists of (potentially multiple-byte) blocks,
+    /// there may be "unused" bits in the final block. We guarantee that, at all points visible to
+    /// the user, the "unused" bits are set to 0.
+    vec: Vec<T>
+}
+
+// In an ideal world, Rust's type defaults would allow us to fold the two impl blocks into one and
+// say "impl Vob<T: usize>", but currently that doesn't work.
+impl Vob<usize> {
```

Comment:
So we are specialising a handful of functions for `usize` backed vobs here? The other functions missing fall back on the generic versions?

Reply:
Correct.